### PR TITLE
GH-35167: [Docs][C++] Use new API for arrow::json::TableReader

### LINKED ISSUE #35167: [Docs] Make signature incorrect in docs for JSON TableReader
### Describe the bug, including details regarding any error messages, version, and platform.

The argument count appears wrong in docs for for the JSON table reader method..

[docs](https://arrow.apache.org/docs/cpp/json.htmlrl): 

```c++
st = arrow::json::TableReader::Make(pool, input, read_options,
parse_options, &reader);
```

object signature in json/reader.cc:

```c++
Result<std::shared_ptr<TableReader>> TableReader::Make(
MemoryPool* pool, std::shared_ptr<io::InputStream> input,
 const ReadOptions& read_options, const ParseOptions& parse_options)
```

It looks correct in the actual API docs.
### Component(s)

C++

### PR DESCRIPTION
**Pull request by rsm-23**
### Rationale for this change

The current document causes a build error.

### What changes are included in this PR?

Use the current API.

### Are these changes tested?

No.

### Are there any user-facing changes?

Yes.

* Closes: #35167

### CODE REVIEW

File: docs/source/cpp/json.rst [hunk 1]
```diff
@@ -59,7 +59,7 @@ the output table.
       // Instantiate TableReader from input stream and options
       std::shared_ptr<arrow::json::TableReader> reader;
       st = arrow::json::TableReader::Make(pool, input, read_options,
```
Comment (wjones127) [inline docs/source/cpp/json.rst:61]: This doesn't return a `Status` anymore, it returns a `Result<std::shared_ptr<arrow::json::TableReader>>`. The CSV docs reflect that by saving it to a variable with a `maybe_` prefix, so we could do the same here:
```suggestion
      auto maybe_reader = arrow::json::TableReader::Make(pool, input, read_options,
```

File: docs/source/cpp/json.rst [hunk 2]
```diff
@@ -59,7 +59,7 @@ the output table.
...
-                                          parse_options, &reader);
+                                          parse_options);
       if (!st.ok()) {
```
Comment (wjones127) [inline docs/source/cpp/json.rst:63]:
```suggestion
      if (!maybe_reader.ok()) {
```

**Review by wjones127 — CHANGES_REQUESTED** [review]
Thanks for submitting a PR. Since the function now uses result instead of an out parameter, we should make it clearer in the example that the return value contains the reader now.

**Comment by rsm-23** [comment]
Thanks for the help! @wjones127

File: docs/source/cpp/json.rst [hunk 3]
```diff
@@ -58,9 +58,9 @@ the output table.
 
       // Instantiate TableReader from input stream and options
       std::shared_ptr<arrow::json::TableReader> reader;
-      st = arrow::json::TableReader::Make(pool, input, read_options,
-                                          parse_options, &reader);
-      if (!st.ok()) {
+      auto maybe_reader = arrow::json::TableReader::Make(pool, input, read_options,
+                                          parse_options);
```
Comment (wjones127) [inline docs/source/cpp/json.rst:62]: Just a quick format and then I think this is good.
```suggestion
      auto maybe_reader =
        arrow::json::TableReader::Make(pool, input, read_options, parse_options);
```

**Comment by rsm-23** [comment]
@wjones127 resolved all the comments.

File: docs/source/cpp/json.rst [hunk 4]
```diff
@@ -58,9 +58,8 @@ the output table.
 
       // Instantiate TableReader from input stream and options
       std::shared_ptr<arrow::json::TableReader> reader;
```
Comment (kou) [inline docs/source/cpp/json.rst:60]: We need to update `arrow::json::TableReader::Read()` too:

```cpp
{
    // ...
    arrow::MemoryPool* pool = default_memory_pool();
    std::shared_ptr<arrow::io::InputStream> input = ...;

    auto read_options = arrow::json::ReadOptions::Defaults();
    auto parse_options = arrow::json::ParseOptions::Defaults();

    // Instantiate TableReader from input stream and options
    auto maybe_reader = arrow::json::TableReader::Make(pool, input, read_options, parse_options);
    if (!maybe_reader.ok()) {
        // Handle TableReader instantiation error...
    }
    auto reader = *maybe_reader;

    // Read table from JSON file
    auto maybe_table = reader->Read();
    if (!maybe_table.ok()) {
        // Handle JSON read error
        // (for example a JSON syntax error or failed type conversion)
    }
    auto table = *maybe_table;
}
```

Points:

* We can remove ` arrow::Status st;`
* We need to extract `reader` from `maybe_reader`
* We need to use `arrow::Result<std::shared_ptr<arrow::Table>> arrow::json::TableReader::Read()`

Comment (rsm-23) [reply docs/source/cpp/json.rst]: Thanks for the points @kou . Made the changes.

**Review by kou — APPROVED** [review]
+1

**Pull request merged**